Migrate Worksheet to DX

Bugerman58 opened the pull request:
## Description of the issue/feature this PR addresses
☝️ NOTE: Please merge #2828  first!

This PR migrates the AT Worksheet to Dexterity for Python 3 compatibility.

## Current behavior before PR

Worksheet is a content type based on the Archetypes framework

## Desired behavior after PR is merged

Worksheet is a content type based on the Dexterity framework

--
I confirm I have tested this PR thoroughly and coded it according to [PEP8][1]
and [Plone's Python styleguide][2] standards.

[1]: https://www.python.org/dev/peps/pep-0008
[2]: https://docs.plone.org/develop/styleguide/python.html

Bugerman58 committed:
Migrate AT Worksheet -> DX

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
add step to upgrade configure

Bugerman58 committed:
fix and refactoring

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
fix linter warning without tests

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
fix lint

Bugerman58 committed:
fix lint and update demodata archive

Bugerman58 committed:
fix lint (

Bugerman58 committed:
bika setup -> senaite registry -> senaite setup; tests fixed.

Bugerman58 committed:
fix lint: remove unused imported

Bugerman58 commented:
Whoo! All test succeed! 🎉
But i have one/two question...

Bugerman58 committed:
fix migrate step

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
fix lint: remove unused imported

Bugerman58 committed:
fix upgrade step for remarks; add widget for remarks field

Bugerman58 commented:
Hi,  @ramonski and @xispa 

Finally, here’s the working migration for Worksheets onboard. Please take a look and share your feedback. The old code was ported as-is, skipping TODOs and optimizations — probably fine for a starting point.

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

ramonski committed:
Merge branch '2.x' into migrate-ws-to-dx

xispa commented:
> Finally, here’s the working migration for Worksheets onboard. Please take a look and share your feedback. The old code was ported as-is, skipping TODOs and optimizations — probably fine for a starting point.Hi, @ramonski and @xispa

Awesome, thanks. Willing to review asap!

Bugerman58 committed:
merge from 2.x; update demodata; fix migrate step for remarks

Bugerman58 committed:
rename package worksheets -> worksheet; optimize imports blocks.

Bugerman58 commented:
> > Finally, here’s the working migration for Worksheets onboard. Please take a look and share your feedback. The old code was ported as-is, skipping TODOs and optimizations — probably fine for a starting point.Hi, @ramonski and @xispa
> 
> Awesome, thanks. Willing to review asap!

Jordi?

xispa commented:
> Jordi?

Yep. On it. Would you mind to resolve the conflicts?

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
merge from 2.x

xispa commented on src/bika/lims/workflow/worksheet/events.py:
```diff
@@ -18,8 +18,55 @@
+]
+
+
+def copy_src_fields_to_dst(src, dst):
```
Please remove this `copy_src_fields_to_dst` function

xispa commented on src/bika/lims/workflow/worksheet/events.py:
```diff
@@ -38,3 +85,149 @@ def after_remove(worksheet):
     api.delete(worksheet, check_permissions=False)
+
+
+def after_reject(worksheet):
```
You can remove this function entirely. At some point, the `workflow_script_reject` function stopped being called, so the worksheet was transitioned to "Rejected" while the analyses remained unchanged. Given that the rejection workflow is unreliable, and the rejection transition hasn't actually worked for a while without anyone noticing, it's better to remove both the `after_reject` handler and the `reject` transition for the `Worksheet` content type.

xispa replied:
Note that removing the rejection workflow also opens the door to removing the `RejectAnalysis` content type. I don't expect recent instances to contain objects of this type, since the worksheet rejection workflow hasn't been functioning for some time, but older instances (including those originally deployed on 1.x versions and upgraded incrementally) may still have them.

Since `RejectAnalysis` objects are contained within `Worksheet` objects, removing all worksheets in the rejected status should, in theory, also remove any existing `RejectAnalysis` objects. The only exception would be cases where a worksheet containing them is in a different status. Therefore, this pull request should also consider removing any remaining `RejectAnalysis` objects and cleaning up worksheets still marked as rejected.

xispa commented on src/senaite/core/browser/modals/sample.py:
```diff
@@ -91,13 +92,14 @@ def create_worksheet_for(self, samples, analyst, categories, template):
 
         # create the new worksheet
         ws = api.create(self.worksheet_folder, "Worksheet")
+        ws.setTitle(ws.getId())
```
Better to  override the function `Title` in the `Worksheet` content type so it always returns the id of the worksheet, like we do for other content types:

```python
@security.protected(permissions.View)
def Title(self):
    return self.getId()
```

Remember to also _flag_ the `title` field as omitted/hidden:

```python
directives.omitted("title")
```

Bugerman58 replied:
> Better to override the function Title in the Worksheet content type so it always returns the id of the worksheet, like we do for other content types:

this has already been done:
https://github.com/Bugerman58/senaite.core/blob/migrate-ws-to-dx/src/senaite/core/content/worksheet.py#L255

`.setTitle` will be removed

xispa commented on src/bika/lims/workflow/worksheet/events.py:
```diff
@@ -38,3 +85,149 @@ def after_remove(worksheet):
+    duplicates and references to match the rejected
+    worksheet.
+    """
+    if skip(worksheet, "reject"):
```
This is the only in all of the `senaite.core` where this function `skip` is used. Please remove `bika.lims.workflow.skip` function along with `after_reject`.

xispa commented on src/bika/lims/workflow/worksheet/events.py:
```diff
@@ -38,3 +85,149 @@ def after_remove(worksheet):
+
+    new_ws.setAnalyses(new_ws_analyses)
+    new_ws.setLayoutView(new_layout)
+    new_ws.setReplacesRejectedWorksheet(worksheet)
```
As per the comment above, is better to remove the `reject` transition. Therefore, this `setReplacesRejectedWorksheet` (and underlying field, if any) can be removed as well.

xispa commented on src/bika/lims/workflow/worksheet/events.py:
```diff
@@ -38,3 +85,149 @@ def after_remove(worksheet):
+    worksheet.REQUEST["context_uid"] = worksheet.UID()
+    worksheet.setLayoutView(old_layout)
+    worksheet.setAnalyses(old_ws_analyses)
+    worksheet.setReplacedBy(new_ws)
```
Likewise, remove the `ReplacedBy` function and underlying field. In addition, remove the function `bika.lims.browser.worksheet.tools.showRejectionMessage` and its calls.

Bugerman58 committed:
fix for printview page

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

xispa commented on src/bika/lims/workflow/analysis/events.py:
```diff
@@ -315,8 +315,9 @@ def remove_analysis_from_worksheet(analysis):
         doActionFor(worksheet, "verify")
-    else:
-        # We've removed all analyses. Rollback to "open"
+    elif api.get_review_status(worksheet) != "open":
```
This condition is not needed. `doActionFor` will attempt the transition but will only succeed if the current status allows `rollback_to_open`, the user has the required permissions, and any associated guard evaluates to `True`. It's better to rely on this built-in workflow machinery rather than adding an extra conditional.

xispa commented on src/bika/lims/content/abstractanalysis.py:
```diff
@@ -1148,7 +1149,7 @@ def getWorksheetUID(self):
         Returns WS UID if this analysis is assigned to a worksheet, or None.
         """
-        uids = get_backreferences(self, relationship="WorksheetAnalysis")
+        uids = get_backrefs(self, relationship="WorksheetAnalysis")
```
Good catch!

Bugerman58 replied:
Thanks

xispa commented on src/bika/lims/content/bikasetup.py:
```diff
@@ -1395,5 +1399,57 @@ def setInvalidationReasonRequired(self, value):
         if setup:
             setup.setInvalidationReasonRequired(value)
 
+    def getWorksheetLayout(self):
```
I didn't see these settings being copied from `bika_setup` to `senaitesetup` in the upgrade step. Please make sure to do so. Otherwise, the settings you moved to setup won't retain their previous values after the upgrade. This applies to the fields `WorksheetLayout`, `RestrictWorksheetUsersAccess`, `RestrictWorksheetManagement`, and `ScientificNotationReport`.

xispa commented on src/bika/lims/content/bikasetup.py:
```diff
@@ -1395,5 +1399,57 @@ def setInvalidationReasonRequired(self, value):
+        if setup:
+            setup.setWorksheetLayout(value)
+
+    def getRestrictWorksheetUsersAccess(self):
```
See comment above. Copy the value of this `WorksheetLayout` field from `bika_setup` to `senaitesetup` in the upgrade step.

xispa commented on src/bika/lims/content/bikasetup.py:
```diff
@@ -1395,5 +1399,57 @@ def setInvalidationReasonRequired(self, value):
+        if setup:
+            setup.setRestrictWorksheetUsersAccess(value)
+
+    def getRestrictWorksheetManagement(self):
```
See comment above. Copy the value of this `RestrictWorksheetManagement` field from `bika_setup` to `setup` in the upgrade step.

xispa commented on src/bika/lims/content/bikasetup.py:
```diff
@@ -1395,5 +1399,57 @@ def setInvalidationReasonRequired(self, value):
+        if setup:
+            setup.setRestrictWorksheetManagement(value)
+
+    def getScientificNotationReport(self):
```
See comment above. Copy the value of this `ScientificNotationReport` field from `bika_setup` to `senaitesetup` in the upgrade step.

xispa commented on src/senaite/core/browser/worksheet/ajax.py:
```diff
@@ -0,0 +1,69 @@
+        value = self.request.get("value", "")
+        if not value:
+            return
+        if not mtool.getMemberById(value):
```
It's better to use `bika.lims.api.user.get_user`, as it correctly handles cases where a user's `id` does not match their `login_name`.:

```python
if not get_user(user=value):
    return
```

xispa commented on src/senaite/core/browser/worksheet/configure.zcml:
```diff
@@ -0,0 +1,136 @@
+      for="senaite.core.interfaces.IWorksheets"
+      name="view"
+      class=".views.WorksheetsView"
+      permission="zope.Public"
```
This should be `permission="cmf.ListFolderContents"` so only users able to list the contents of `worksheets` folder have access granted to this view.

xispa commented on src/senaite/core/browser/worksheet/configure.zcml:
```diff
@@ -0,0 +1,136 @@
+      for="senaite.core.interfaces.IWorksheet"
+      name="view"
+      class=".views.WorksheetView"
+      permission="zope.Public"
```
This should be `permission="zope2.View"` instead. Only authenticated users should have access. See https://5.docs.plone.org/develop/plone/security/standard_permissions.html#standard-permissions

xispa commented on src/senaite/core/browser/worksheet/configure.zcml:
```diff
@@ -0,0 +1,136 @@
+      for="senaite.core.interfaces.IWorksheets"
+      name="worksheet_add"
+      class=".views.AddWorksheetView"
+      permission="zope.Public"
```
Permission should be `senaite.core.permissions.worksheet.AddWorksheet`

xispa commented on src/senaite/core/browser/worksheet/views/add_analyses.py:
```diff
@@ -0,0 +1,332 @@
+
+        return self.template()
+
+    def rejection_message(self):
```
Remove this. This is no longer required, see my comment above regarding rejected worksheets.

xispa commented on src/senaite/core/browser/worksheet/views/add_analyses.py:
```diff
@@ -0,0 +1,332 @@
+    def __call__(self):
+        super(AddAnalysesView, self).__call__()
+
+        self.rejection_message()
```
Remove

xispa commented on src/senaite/core/browser/worksheet/views/add_analyses.py:
```diff
@@ -0,0 +1,332 @@
+    def getWorksheetTemplates(self):
+        """Return WS Templates
+        """
+        vocabulary = CatalogVocabulary(self)
```
Pheeeeew… if this `CatalogVocabulary` works, it does so by accident. For example, note that [this line](https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/vocabularies/__init__.py#L78) calls `brain.getObjec()`, which will never work.

As far as I can tell, `getWorksheetTemplates` is only used in [`add_analyses.pt`](https://github.com/senaite/senaite.core/pull/2805/files#diff-6f45ac024f58f98904166c7a61a4c2aed1043488980640f87af61522ffe300b7). Instead of the current approach, simply search the catalog for active `WorksheetTemplate` and return a list of tuples like `(<uid>, <title>)`. This has the added benefit that you won't need to wake up the objects, since the `brain` already contains both `UID` and `Title` as metadata.

Bugerman58 replied:
> Pheeeeew… if this CatalogVocabulary works, it does so by accident. For example, note that [this line](https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/vocabularies/__init__.py#L78) calls brain.getObjec(), which will never work.

)) This code was copied from here: https://github.com/senaite/senaite.core/blob/2.x/src/bika/lims/browser/worksheet/views/add_analyses.py#L301

Yes, of course I'll fix it

xispa commented on src/senaite/core/browser/worksheet/views/add_duplicate.py:
```diff
@@ -0,0 +1,206 @@
+    def __call__(self):
+        template = super(AddDuplicateView, self).__call__()
+
+        self.rejection_message()
```
Remove

xispa commented on src/senaite/core/browser/worksheet/views/add_duplicate.py:
```diff
@@ -0,0 +1,206 @@
+            self.handle_submit()
+        return template
+
+    def rejection_message(self):
```
Remove this function. See my comment regarding to rejected Worksheets above

xispa commented on src/senaite/core/browser/worksheet/views/add_worksheet.py:
```diff
@@ -0,0 +1,89 @@
+        portal = api.get_portal()
+        ws_container = portal.get(WORKSHEETS_FOLDER_ID)
+        ws = api.create(ws_container, "Worksheet")
+        ws.setTitle(ws.getId())
```
As mentioned before, better to override the `Title` function in `Worksheet` content type to return the `id` directly

xispa commented on src/senaite/core/browser/worksheets/view.py:
```diff
@@ -0,0 +1,476 @@
+from senaite.core.permissions.worksheet import can_manage_worksheets
+
+
+class WorksheetsView(ListingView):
```
To follow the latest conventions, it's better to rename `folder.py` to `view.py` and move it under the `senaite.core.browser.worksheets` namespace (note the plural worksheet**s**). Then, create a separate module `senaite.core.browser.worksheets.worksheet` to contain all worksheet-specific views and related code.

You can pick the structure of `senaite.core.browsers.clients` as an example.

xispa commented on src/senaite/core/browser/worksheet/views/manage_results.py:
```diff
@@ -0,0 +1,226 @@
+    def __call__(self):
+        self.update()
+
+        self.rejection_message()
```
Remove

xispa commented on src/senaite/core/browser/worksheet/views/manage_results.py:
```diff
@@ -0,0 +1,226 @@
+        layout_view.before_render()
+        return layout_view.contents_table()
+
+    def rejection_message(self):
```
Remove the function `rejection_message`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+            return wst.absolute_url_path()
+        return ""
+
+    def getRejectionMessages(self):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        mutator(self, value)
+
+    @security.protected(permissions.View)
+    def getReplacesRejectedWorksheet(self):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        accessor = self.accessor("replaces_rejected_worksheet")
+        return accessor(self)
+
+    def getRawReplacesRejectedWorksheet(self):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        return None
+
+    @security.protected(permissions.ModifyPortalContent)
+    def setReplacesRejectedWorksheet(self, value):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        accessor = self.accessor("replaced_by")
+        return accessor(self)
+
+    def getRawReplacedBy(self):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+    ResultsLayout = property(getResultsLayout, setResultsLayout)
+
+    @security.protected(permissions.View)
+    def getReplacedBy(self):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        default=DEFAULT_WORKSHEET_LAYOUT,
+    )
+
+    replaced_by = UIDReferenceField(
```
Remove this field. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        required=False,
+    )
+
+    replaces_rejected_worksheet = UIDReferenceField(
```
Remove this field. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+            default=u"Method"
+        ),
+        allowed_types=("Instrument",),
+        vocabulary="senaite.core.vocabularies.available_methods",
```
This vocabulary is wrong

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        title=_(
+            u"title_worksheet_method",
+            default=u"Method"
+        ),
```
This is wrong, probably a copy&paste of `method`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+    )
+
+    # TODO Remove. Instruments must be assigned directly to each analysis.
+    instrument = UIDReferenceField(
```
You could use this Pull Request as an opportunity to remove the instrument field entirely. The intended behavior is that the instrument selector remains available in the Worksheet view, but when an instrument is selected, the system should assign it individually to each analysis in the worksheet that supports that instrument.

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+            default=u"Analyses"
+        ),
+        allowed_types=("Analysis", "DuplicateAnalysis", "ReferenceAnalysis",
+                       "RejectAnalysis"),
```
Remove `RejectAnalysis`, see my comment regarding to the rejection workflow

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+    """Worksheet schema interface
+    """
+
+    title = schema.TextLine(
```
Add `directives.omitted("title")`, cause we don't want this field to be editable.

xispa replied:
It's better to use `senaite.core.schema.TextLineField` instead of `zope.schema.TextLine` whenever possible, as it automatically handles `unicode` and `utf-8`, preserving compatibility with the legacy behavior of auto-generated setters and getters in AT content types.

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        required=True,
+    )
+
+    description = schema.Text(
```
Add `directives.omitted("description")`. For now, this field is only here for convenience

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        required=False,
+    )
+
+    analyst = schema.TextLine(
```
Better use `senaite.core.schema.TextLineField`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        return None
+
+    @security.protected(permissions.ModifyPortalContent)
+    def setReplacedBy(self, value):
```
Remove this function. It is no longer necessary after the removal of the `reject` transition and related contents. See comment for `after_reject`

xispa commented on src/senaite/core/content/worksheet.py:
```diff
@@ -0,0 +1,1556 @@
+        mutator(self, sorted_layout)
+
+    # BBB: AT schema field property
+    # Layout = property(getLayoutView, setLayoutView)
```
Please retain the `getLayout` and `setLayout` methods to ensure backwards compatibility. Keep in mind that other add-ons might rely on these functions (since you renamed the original `Layout` field to `layout_view`), and we don't want to break them. Also, uncomment the `Layout` property for the same reason. You should add a `LayoutView` property as well.

Alternatively, you could rename the field `layout_view` back to `layout` and update its accessor and mutator accordingly. I would prefer this approach, as the name `LayoutView` can be misleading—it might suggest a browser view (leading one to expect a `BrowserView` object), whereas this field actually stores the positions/slots of each analysis in the worksheet.

xispa commented on src/senaite/core/exportimport/instruments/foss/fiastar/fiastar.py:
```diff
@@ -63,7 +63,7 @@ def __call__(self, analyses):
         # for looking up "cup" number (= slot) of ARs
         parent_to_slot = {}
-        layout = self.context.getLayout()
+        layout = self.context.getLayoutView()
```
Note my comment regarding AT's `Layout` vs DX's `layout_view  field in `Worksheet`

Bugerman58 replied:
Hi, @xispa 
If we leave the getter unchanged(`getLayout`), an error occurs when switching to the worksheet view:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 264, in publish
  Module ZPublisher.BaseRequest, line 450, in traverse
  Module plone.rest.patches, line 15, in __before_publishing_traverse__
  Module Products.CMFCore.DynamicType, line 164, in __before_publishing_traverse__
  Module Products.CMFDynamicViewFTI.fti, line 239, in queryMethodID
  Module Products.CMFDynamicViewFTI.fti, line 201, in defaultView
  Module Products.CMFPlone.PloneTool, line 696, in browserDefault
TypeError: getLayout() got an unexpected keyword argument 'check_exists'
```

For this reason, the field was renamed(`getLayout` -> `getTemplateLayout`) in a `WorksheetTemplate` content type: https://github.com/senaite/senaite.core/pull/2599#issuecomment-2376098725

https://github.com/plone/Products.CMFPlone/blob/5.2.x/Products/CMFPlone/PloneTool.py#L702

xispa commented on src/senaite/core/exportimport/instruments/lachat/quickchem.py:
```diff
@@ -208,7 +208,7 @@ def __call__(self, analyses):
         # Column D empty
         # Column E should always be 1 (2 indicates a duplicate from the same cup)
-        layout = self.context.getLayout()
+        layout = self.context.getLayoutView()
```
Note my comment regarding AT's `Layout` vs DX's `layout_view  field in `Worksheet`

Bugerman58 replied:
[comment above](https://github.com/senaite/senaite.core/pull/2805#discussion_r2605251285)

xispa commented on src/senaite/core/exportimport/instruments/varian/vistapro/icp.py:
```diff
@@ -212,7 +212,7 @@ def __call__(self, analyses):
         # If Reference analysis, COL B is the Ref Sample.
         # If Duplicate analysis, COL B is the Worksheet.
-        lyt = self.context.getLayout()
+        lyt = self.context.getLayoutView()
```
Note my comment regarding AT's `Layout` vs DX's `layout_view  field in `Worksheet`

Bugerman58 replied:
[comment above](https://github.com/senaite/senaite.core/pull/2805#discussion_r2605251285)

xispa commented on src/senaite/core/browser/worksheet/views/referencesamples.py:
```diff
@@ -0,0 +1,271 @@
+        ]
+
+    def __call__(self):
+        self.rejection_message()
```
remove

xispa commented on src/senaite/core/browser/worksheet/views/referencesamples.py:
```diff
@@ -0,0 +1,271 @@
+            self.handle_submit()
+        return template
+
+    def rejection_message(self):
```
remove

xispa commented on src/senaite/core/browser/worksheets/worksheet/view.py:
```diff
@@ -0,0 +1,43 @@
+
+    def __call__(self):
+        view = "add_analyses"
+        if self.context.getLayoutView():
```
Note my comment regarding AT's `Layout` vs DX's `layout_view` field in `Worksheet`. Specially relevant here that fact the name can lead to confusion `BrowserView` vs list of dicts

xispa commented on src/senaite/core/interfaces/__init__.py:
```diff
@@ -511,6 +511,26 @@ class ISampleTypes(Interface):
+    """Marker interface for additional Worksheet layouts
+    """
+
+    def getLayouts(self):
```
Might be worth to rename this to `ResultLayouts` to better resemble with the field `result_layout` (`ResultLayout`) from `Worksheet`.

xispa commented on src/senaite/core/upgrade/v02_07_000.py:
```diff
@@ -551,3 +583,167 @@ def setup_custom_image_and_file_types(tool):
+    origin = api.get_portal().worksheets
+
+    # get the destination container
+    temp_folder = get_destination_folder("worksheets-tmp")
```
I would approach this the other way around. Temporarily rename the existing worksheets folder to `worksheets_old`, then create the new DX folder `worksheets` and add all the objects there. Keep in mind that paths are indexed in the catalog, so if you create all the worksheets in the temporary folder and rename it later, you would need to reindex every worksheet to ensure the path is correctly stored in both the `worksheets_catalog` and the `uid_catalog`.

xispa commented on src/senaite/core/upgrade/v02_07_000.py:
```diff
@@ -551,3 +583,167 @@ def setup_custom_image_and_file_types(tool):
+
+    # get the new destination container for DX Worksheet folder
+    new_folder = get_destination_folder("worksheets")
+    for ws in temp_folder.objectValues():
```
I see, so you're first creating the worksheet in the temp folder and then moving it to the final folder. I think it's better to avoid reindexing the new Worksheet objects twice. First, rename the old folder (without notifying the children), and then create the final DX folder, where you can directly create the new `Worksheet` objects.

Bugerman58 replied:
Hi, @xispa 
> First, rename the old folder...

I followed your advice and tried renaming the folder at the very beginning of the upgrade step, but I get this stack trace:
```shell
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module Products.GenericSetup.tool, line 1135, in manage_doUpgrades
  Module Products.GenericSetup.upgrade, line 185, in doStep
  Module senaite.core.upgrade, line 39, in wrap_func_args
  Module senaite.core.upgrade.v02_07_000, line 597, in migrate_worksheets_to_dx
  Module Products.CMFPlone.PloneFolder, line 102, in manage_renameObject
  Module OFS.CopySupport, line 371, in manage_renameObject
  Module Products.CMFCore.PortalFolder, line 418, in _verifyObjectPaste
ValueError: Disallowed subobject type: WorksheetFolder
```
This might be due to the missing `WorksheetFolder.xml`, as it's already been moved to the DX.

xispa replied:
It should be possible before you flush the `Worksheet` portal type from `portal_types` (before the call to `remove_at_portal_types ` function). You need first to allow the creation of objects from type `WorksheetFolder` inside `portal` though. Something like:

```python
# get the current allowed types for portal
portal = api.get_portal()
pt = api.get_tool("portal_types")
ti = pt.getTypeInfo(portal)
allowed_types = ti.allowed_content_types

# temporarily append WorksheetFolder as an allowed type
portal_type = "WorksheetFolder"
if portal_type not in ti.allowed_content_types:
    ti.allowed_content_types = allowed_types + (portal_type, )

# Rename the worksheets folder here
...

# restore the allowed_types
ti.allowed_content_types = allowed_types

# ensure old AT types are flushed first
remove_at_portal_types(tool)

# run required import steps
tool.runImportStepFromProfile(profile, "typeinfo")
...
```

Bugerman58 replied:
Oh, thanks for the tip. I'll try that.

xispa commented on src/senaite/core/upgrade/v02_07_000.zcml:
```diff
@@ -2,6 +2,14 @@
+      destination="2716"
+      handler=".v02_07_000.migrate_worksheets_to_dx"
+      profile="senaite.core:default"/>
+
```
Remember to add another upgrade step (or maybe in the same) to _port_ the values of the settings related with worksheets from `bikla_setup` to `senaitesetup`.

xispa commented on src/senaite/core/z3cform/widgets/remarks/__init__.py:
```diff
@@ -0,0 +1,21 @@
+# Copyright 2018-2025 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+from .widget import RemarksWidget
```
How about isolating the `RemarksField` + `RemarksWidget` addition in a separate Pull Request?

Bugerman58 replied:
#2828

xispa commented on src/senaite/core/schema/__init__.py:
```diff
@@ -57,6 +59,7 @@
 classImplementsFirst(IntField, IIntField)
 classImplementsFirst(MultiUploadField, IMultiUploadField)
 classImplementsFirst(PhoneField, IPhoneField)
+classImplementsFirst(RemarksField, IRemarksField)
```
How about isolating the `RemarksField` + `RemarksWidget` addition in a separate Pull Request?

Bugerman58 replied:
#2828

xispa reviewed (changes requested):
Thanks @Bugerman58 for your huge effort here!

I have left some messages. Also, after the upgrade I get the following traceback when I try to see an existing worksheet:

```
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.core.browser.worksheet.views.manage_results, line 47, in __call__
  Module senaite.core.browser.worksheet.views.manage_results, line 89, in update
  Module plone.autoform.view, line 29, in update
  Module plone.autoform.view, line 49, in _update
  Module plone.autoform.base, line 67, in updateFieldsFromSchemata
  Module plone.dexterity.browser.view, line 15, in schema
  Module zope.component._api, line 165, in getUtility
ComponentLookupError: (<InterfaceClass plone.dexterity.interfaces.IDexterityFTI>, 'Worksheet')
```

Bugerman58 committed:
intermediate commit after review

Bugerman58 committed:
fix permission

Bugerman58 committed:
fix check get_user

Bugerman58 committed:
fix vocabulary for instrument field

Bugerman58 committed:
copy settings from bika_setup to senaitesetup into upgrade step

Bugerman58 committed:
change structure folders

Bugerman58 committed:
fix import views

Bugerman58 committed:
fix include package into configure

Bugerman58 committed:
change template for folder view

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
fix after review

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
remove unused imports

Bugerman58 committed:
fix test for removed workflow status 'rejected' for ws

Bugerman58 committed:
change condition for select view

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
change upgrade step: first rename old folder

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
fix worksheet layouts vocabulary

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
merge from 2.x

Bugerman58 committed:
remove unused imports

xispa commented:
Thanks for https://github.com/senaite/senaite.core/pull/2828 . Will resume the review of this one!

Bugerman58 committed:
Merge branch '2.x' into migrate-ws-to-dx

xispa committed:
Fix TypeError: getAnalyses() got an unexpected keyword argument 'full_objects'

Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 176, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 385, in publish_module
  Module ZPublisher.WSGIPublisher, line 288, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 63, in call_object
  Module senaite.core.browser.worksheets.worksheet.manage_results, line 84, in __call__
  Module Products.Five.browser.pagetemplatefile, line 126, in __call__
  Module Products.Five.browser.pagetemplatefile, line 61, in __call__
  Module zope.pagetemplate.pagetemplate, line 135, in pt_render
  Module Products.PageTemplates.engine, line 378, in __call__
  Module z3c.pt.pagetemplate, line 176, in render
  Module chameleon.zpt.template, line 302, in render
  Module chameleon.template, line 215, in render
  Module chameleon.template, line 192, in render
  Module 1ae0dde3615a24c131608fdf7fab4250, line 1816, in render
  Module 3b4abe6abe95c5b8cf33b79d89b266cc, line 1428, in render_master
  Module 3b4abe6abe95c5b8cf33b79d89b266cc, line 407, in render_content
  Module 1ae0dde3615a24c131608fdf7fab4250, line 1709, in __fill_content_core
  Module zope.tales.pythonexpr, line 73, in __call__
   - __traceback_info__: (view.contents_table())
  Module <string>, line 1, in <module>
  Module senaite.core.browser.worksheets.worksheet.manage_results, line 94, in contents_table
  Module senaite.core.browser.worksheets.worksheet.analyses_listing, line 150, in before_render
  Module senaite.core.browser.worksheets.worksheet.analyses_listing, line 165, in enable_remarks
  Module senaite.core.browser.worksheets.worksheet.analyses_listing, line 201, in show_analysis_remarks_transition
  Module senaite.core.browser.worksheets.worksheet.analyses_listing, line 211, in get_analyses
TypeError: getAnalyses() got an unexpected keyword argument 'full_objects'

xispa committed:
Merge branch '2.x' of github.com:senaite/senaite.core into Bugerman58-migrate-ws-to-dx

xispa reviewed (approved):
Excellent work, thanks!

Bugerman58 commented:
> Excellent work, thanks!

Thanks, Jordi!

ramonski commented:
It seems like worksheet templates do no longer apply the layout and analyses automatically to the configured slots.

Bugerman58 commented:
> It seems like worksheet templates do no longer apply the layout and analyses automatically to the configured slots.

Hi Ramon!

I created a test template and applied it using the control in the newly created worksheet. After applying the template, both the tests and their duplicates were arranged correctly according to the defined layout scheme.

WorskheetTemplate:

<img width="592" height="505" alt="Screenshot 2026-03-02 at 15 40 22" src="https://github.com/user-attachments/assets/5477658a-b256-491d-bbcc-6fb1e7706f97" />


Worksheet after apply template:

<img width="597" height="612" alt="Screenshot 2026-03-02 at 15 41 31" src="https://github.com/user-attachments/assets/5b3d737e-942e-45f4-a838-9ba0779e0ce2" />

ramonski commented:
Thanks for your quick response Anton!
I just see now that I had the restrict to Method/Instrument active. I guess that this was the reason it skipped the analyses.
I will check this in some more detail and report back if something unexpected happens.
Thanks again!
